File: Python/Minimum Path Sum.py
# Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right which minimizes the sum of all numbers along its path.

# Note: You can only move either down or right at any point in time.

class Solution:
    # @param grid, a list of lists of integers
    # @return an integer
    def minPathSum(self, grid):
		if len(grid) == 0:
			return 0
		m = len(grid)
		n = len(grid[0])
		MinV = [[0 for i in xrange(n)] for j in xrange(m)]
		MinV[0][0] = grid[0][0]
		for i in range(1,m):
			MinV[i][0] = MinV[i-1][0] + grid[i][0]
		for i in range(1,n):
			MinV[0][i] = MinV[0][i-1] + grid[0][i]
		for i in range(1,m):
			for j in range(1,n):
				MinV[i][j] = min(MinV[i-1][j],MinV[i][j-1]) + grid[i][j]
		return MinV[m-1][n-1]

File: Java/Minimum Path Sum.java
class Minimum Path Sum{
	public int minPathSum(int[][] grid) {
        if(grid == null || grid.length == 0)
        	return 0;
        int rowLen = grid.length;
        int colLen = grid[0].length;
        int[][] min = new int[rowLen][colLen];
        min[0][0] = grid[0][0];
        for(int i=1;i<colLen;i++){
        	min[0][i] = min[0][i-1] + grid[0][i];
        }
        for(int i=1;i<rowLen;i++){
        	min[i][0] = min[i-1][0] + grid[i][0];
        }
        for(int i=1;i<rowLen;i++){
        	for(int j=1;j<colLen;j++){
        		min[i][j] = grid[i][j] + Math.min(min[i-1][j],min[i][j-1]);
        	}
        }
        return min[rowLen-1][colLen-1];
    }
}